packages/client/lib/sync/txpool.ts

```diff
@@ -419,6 +427,92 @@ export class TxPool {
     })
   }
 
+  /**
+   * Helper to return a normalized gas price across different transaction types.
+   * @param tx The tx
+   * @param priority If a 1559 tx and true, returns the maxPriorityFeePerGas. If false, returns the maxFeePerGas. (default: true)
+   */
+  private txGasPrice(tx: TypedTransaction, priority = true) {
+    if (tx.supports(Capability.EIP1559FeeMarket)) {
+      if (priority) {
+        return (tx as FeeMarketEIP1559Transaction).maxPriorityFeePerGas
+      } else {
+        return (tx as FeeMarketEIP1559Transaction).maxFeePerGas
+      }
+    } else {
+      return (tx as Transaction).gasPrice
+    }
+  }
+
+  /**
+   * Returns eligible txs to be mined sorted by price in such a way that the
+   * nonce orderings within a single account are maintained.
+   *
+   * Note, this is not as trivial as it seems from the first look as there are three
+   * different criteria that need to be taken into account (price, nonce, account
+   * match), which cannot be done with any plain sorting method, as certain items
+   * cannot be compared without context.
+   *
+   * This method first sorts the separates the list of transactions into individual
+   * sender accounts and sorts them by nonce. After the account nonce ordering is
+   * satisfied, the results are merged back together by price, always comparing only
+   * the head transaction from each account. This is done via a heap to keep it fast.
+   *
+   * @param stateManager Account nonces are queried to only include executable txs
+   * @param baseFee Provide a baseFee to exclude txs with a lower gasPrice
+   */
+  async txsByPriceAndNonce(stateManager: StateManager, baseFee?: BN) {
+    const txs: TypedTransaction[] = []
+    // Separate the transactions by account and sort by nonce
+    const byNonce = new Map<string, TypedTransaction[]>()
+    for (const [address, poolObjects] of this.pool) {
+      let txsSortedByNonce = poolObjects
+        .map((obj) => obj.tx)
+        .sort((a, b) => a.nonce.sub(b.nonce).toNumber())
```

Comment:
I guess we might want to directly have this in the pool data structure itself that the txs are sorted by nonce when added? 🤔 (I somehow assumed that this would "naturally" be the case by the order they are added but - you are right - this is likely not always the case).

Reply:
I think it might be more work than needed right now to maintain the order on insertion/deletion, as it might be fragile since txs can be replaced/repriced. The sort here should be relatively quick and saves headaches in other places.